apps/testingaccessibility/src/server/ability.ts

```diff
@@ -1,64 +1,73 @@
-import {AbilityBuilder, Ability} from '@casl/ability'
-import {isNull, isUndefined, some} from 'lodash'
-import {Coupon, Purchase} from '@prisma/client'
+import {AbilityBuilder, Ability, AbilityClass} from '@casl/ability'
+import {isEmpty} from 'lodash'
+import {
+  hasAvailableSeats,
+  hasBulkPurchase,
+  hasValidPurchase,
+} from '../utils/purchase-validators'
 
 type Actions = 'manage' | 'invite' | 'view'
 type Subjects = 'Team' | 'Purchase' | 'Content' | 'Product' | 'all'
 type AppAbility = Ability<[Actions, Subjects]>
+const AppAbility = Ability as AbilityClass<AppAbility>
 
-export async function getAbilityFromToken(token: any) {
-  return defineAbilityFor(token)
+type ViewerAbilityInput = {
+  purchases?: any[]
+  rules?: any
 }
 
-export function bulkCouponHasSeats(coupon: Coupon) {
-  return coupon.usedCount < coupon.maxUses
-}
-
-export function hasBulkPurchase(purchases?: any[]) {
-  return some(purchases, (purchase) => {
-    return !isNull(purchase.bulkCoupon)
-  })
-}
-
-export function hasAvailableSeats(purchases?: any[]) {
-  return some(purchases, (purchase) => {
-    return (
-      !isNull(purchase.bulkCoupon) && bulkCouponHasSeats(purchase.bulkCoupon)
-    )
-  })
-}
-
-export function hasValidPurchase(purchases?: any[]) {
-  return some(purchases, (purchase) => {
-    return isNull(purchase.bulkCoupon) || isUndefined(purchase.bulkCoupon)
-  })
+/**
+ * Creates an ability uses cached rules or the properties
+ * to define a set of rules (purchases)
+ * @param viewerAbilityInput
+ */
+export function getCurrentAbility(
+  viewerAbilityInput: ViewerAbilityInput,
+): AppAbility {
+  return defineAbilityFor(viewerAbilityInput)
 }
 
 /**
+ * factory to create an instance of an app ability
+ * to use and determine if the viewer has access to
+ * various things
+ *
+ * if the input has a set of rules, use those rules
+ * otherwise, use the purchases and create rules
  * @see {@link https://casl.js.org/v5/en/guide/define-rules#ability-builder-class|AbilityBuilder}
- * @param token the JWT session token from next-auth with purchases
+ * @param viewerAbilityInput
  */
-function defineAbilityFor(token: any) {
-  const {can, build} = new AbilityBuilder<AppAbility>(Ability)
+export function defineAbilityFor(viewerAbilityInput: ViewerAbilityInput) {
+  const rules = isEmpty(viewerAbilityInput.rules)
+    ? defineRulesForPurchases(viewerAbilityInput.purchases || [])
+    : viewerAbilityInput.rules
+  return new AppAbility(rules)
+}
 
-  console.log({token})
+/**
+ * Creates a structure of rules to use in the ability from list of purchases
+ *
+ * @param purchases
+ */
+export function defineRulesForPurchases(purchases: any[]) {
```

Comment:
this can be called directly with purchases array to get a rules abject that can be cached and used later